src/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.ts

```diff
@@ -382,13 +422,62 @@ g.test('source_imageBitmap,state')
     );
   });
 
+g.test('source_canvas,state')
+  .desc(
+    `
+  Test HTMLCanvasElement as source image in state [placeholder, valid].
+
+  Call 'transferControlToOffscreen' on HTMLCanvasElement will cause the
+  canvas control right transfer. And this canvas is in state 'placeholder'
+  
+  Check whether 'InvalidStateError' is generated when HTMLCanvasElement is
+  in 'placeholder' state.
+    `
+  )
+  .params(u =>
+    u //
+      .combine('state', ['placeholder', 'valid'])
+      .beginSubcases()
+      .combine('copySize', [
+        { width: 0, height: 0, depthOrArrayLayers: 0 },
+        { width: 1, height: 1, depthOrArrayLayers: 1 },
+      ])
+  )
+  .fn(async t => {
+    const { state, copySize } = t.params;
+    const canvas = createOnscreenCanvas(t, 1, 1);
+    if (typeof canvas.transferControlToOffscreen === 'undefined') {
+      t.skip("Browser doesn't support HTMLCanvasElement transfer control right");
+      return;
+    }
+
+    const dstTexture = t.device.createTexture({
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      format: 'bgra8unorm',
+      usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    if (state === 'placeholder') {
+      canvas.transferControlToOffscreen();
+    } else {
+      assert(canvas.getContext('2d') !== null);
```

Comment:
Just checking, do we have any tests that try to use a canvas with _no_ context as a copy source?

There's a little bit too much different between the "placeholder" and "valid" cases. The placeholder case doesn't create a context at all. So it's possible that the "placeholder" case could be rejected by the browser for the wrong reason, and the test would still pass.

Perhaps adding two additional cases would solve this:
- One where the canvas is not transferred AND it has no context
- One where the canvas is transferred, and the OffscreenCanvas has a context